### PR TITLE
Migrate make to ICX/ICPX for Windows

### PR DESCRIPTION
**Pull request by Alexandr-Solovev**
# Description
In this pull request, we are migrating the project's default build system on Windows from the ICL to using ICX.
All changes are necessary to build oneDAL properly.
`DAAL_EXPORT` was removed from some classes, due to correct dllexport build. In general just called functions should be marked as DAAL_EXPORT, and derived classes shouldn't. Most of the changes were done to have green CI.
The concatenate operation `##` works pretty strange and MKL function macros have been updated for it.
`[[maybe_unused]] ` is necessary for avoid compiler issues.
The base version of C++ has been increased to C++17.

### CODE REVIEW

File: .ci/scripts/describe_system.sh [hunk 1]
```diff
@@ -54,6 +54,10 @@ if [ -x "$(command -v icc)" ]; then
     echo "ICC:"
     icc --version
 fi
+if [ -x "$(command -v icx)" ]; then
+    echo "ICX:"
+    icx --version
+fi
```
Comment (Alexsandruss) [inline .ci/scripts/describe_system.sh:60]: AR: same should be applied for sklearnex repo script

**Review by Alexsandruss — CHANGES_REQUESTED** [review]
Major binary version of library should be increased: these changes break backward binary compatibility.

File: cpp/daal/src/externals/core_threading_win_dll.cpp [hunk 2]
```diff
@@ -41,7 +41,11 @@ daal::services::Environment::LibraryThreadingType __daal_serv_get_thr_set();
 
 DAAL_EXPORT HMODULE load_onedal_thread_dll()
 {
-    return DAAL_LOAD_DLL(_DLL_SUFFIX("onedal_thread"));
```
Comment (ethanglaser) [inline cpp/daal/src/externals/core_threading_win_dll.cpp:44]: why is this no longer working?

File: dev/make/compiler_definitions/dpcpp.mk [hunk 3]
```diff
@@ -28,7 +28,7 @@ CORE.SERV.COMPILER.dpcpp = generic
 -Zl.dpcpp =
 -DEBC.dpcpp = -g
 
-COMPILER.lnx.dpcpp = icpx -fsycl -m64 -stdlib=libstdc++ -fgnu-runtime -fwrapv \
+COMPILER.lnx.dpcpp = icpx -fsycl -O0 -m64 -stdlib=libstdc++ -fgnu-runtime -fwrapv \
```
Comment (ethanglaser) [inline dev/make/compiler_definitions/dpcpp.mk:31]: what is this flag?

File: cpp/daal/src/services/service_defines.h [hunk 4]
```diff
@@ -71,7 +71,7 @@ DAAL_EXPORT bool daal_check_is_intel_cpu();
     #define PRAGMA_ICC_TO_STR(ARGS)
     #define PRAGMA_ICC_OMP(ARGS)
     #define PRAGMA_ICC_NO16(ARGS)
-    #define DAAL_TYPENAME
+    #define DAAL_TYPENAME typename
```
Comment (ethanglaser) [inline cpp/daal/src/services/service_defines.h:74]: what does this do?

File: dev/make/compiler_definitions/icx.mkl.32e.mk [hunk 5]
```diff
@@ -18,29 +18,40 @@
 #  Intel compiler definitions for makefile
 #--
 
-PLATs.icx = lnx32e mac32e
+PLATs.icx = lnx32e win32e
 
-CMPLRDIRSUFF.icx = _icx
+CMPLRDIRSUFF.icx =
```
Comment (ethanglaser) [inline dev/make/compiler_definitions/icx.mkl.32e.mk:23]: should it be empty?

File: dev/make/compiler_definitions/icx.mkl.32e.mk [hunk 6]
```diff
@@ -18,29 +18,40 @@
 #  Intel compiler definitions for makefile
 #--
 
-PLATs.icx = lnx32e mac32e
```
Comment (ethanglaser) [inline dev/make/compiler_definitions/icx.mkl.32e.mk:21]: should mac32e be deleted? or just windows added?

**Review by ethanglaser — COMMENTED** [review]
Overall looks good, just have a few clarification questions. Any perf results?

File: cpp/daal/src/externals/core_threading_win_dll.cpp (hunk 2 above)
Comment (Alexandr-Solovev) [reply cpp/daal/src/externals/core_threading_win_dll.cpp:44]: Idk, but in ICX the concatenate operation ## works pretty strange and MKL function macros have been updated for it.

File: dev/make/compiler_definitions/dpcpp.mk (hunk 3 above)
Comment (Alexandr-Solovev) [reply dev/make/compiler_definitions/dpcpp.mk:31]: Now, I am mostly experimenting with different compiler flags for optimization and try to found the best performance option, will deleted if there is no success

File: cpp/daal/src/services/service_defines.h (hunk 4 above)
Comment (Alexandr-Solovev) [reply cpp/daal/src/services/service_defines.h:74]: Add typename instead DAAL_TYPENAME for some classes, it's necessary for icx and doesnt affect icc

File: dev/make/compiler_definitions/icx.mkl.32e.mk (hunk 5 above)
Comment (Alexandr-Solovev) [reply dev/make/compiler_definitions/icx.mkl.32e.mk:23]: Its the question how our release folder should look: __release_lnx or __release_lnx_icx

File: dev/make/compiler_definitions/icx.mkl.32e.mk (hunk 6 above)
Comment (Alexandr-Solovev) [reply dev/make/compiler_definitions/icx.mkl.32e.mk:21]: YEs, we dont have support icx on macos

File: dev/make/compiler_definitions/dpcpp.mk (hunk 3 above)
Comment (ethanglaser) [reply dev/make/compiler_definitions/dpcpp.mk:31]: any details on what the flag actually does? and how much impact does it have on perf?

File: cpp/daal/src/externals/core_threading_win_dll.cpp (hunk 2 above)
Comment (ethanglaser) [reply cpp/daal/src/externals/core_threading_win_dll.cpp:44]: I'd recommend reverting changes to this if possible - it seems like the original should be working (just this deletion and 4 lines of additions)

File: dev/make/compiler_definitions/dpcpp.mk (hunk 3 above)
Comment (napetrov) [reply dev/make/compiler_definitions/dpcpp.mk:31]: this is optimization level flag - fine for experiments but would have implications on perf and library
https://gcc.gnu.org/onlinedocs/gcc/Optimize-Options.html

File: cpp/daal/src/externals/core_threading_win_dll.cpp (hunk 2 above)
Comment (Alexandr-Solovev) [reply cpp/daal/src/externals/core_threading_win_dll.cpp:44]: It doesnt work

File: dev/make/compiler_definitions/dpcpp.mk (hunk 3 above)
Comment (ethanglaser) [reply dev/make/compiler_definitions/dpcpp.mk:31]: Seems like -O0 is default? Are perf results consistent or could differences be variable

**Comment by ethanglaser** [comment]
Any idea on windows examples internal CI fails?

**Comment by samir-nasibli** [comment]
@Alexandr-Solovev 
please update the description for the internal CI fails and actions required for them.

**Review by ethanglaser — APPROVED** [review]
LGTM, mostly just a question on when it makes sense to merge it. Would it make sense to validate on nightly scope? and would it make sense to check perf?

**Comment by Alexandr-Solovev** [comment]
@ethanglaser Thanks for the review! I guess @napetrov knows better when we can merge it? I will launch nightly job + benchmarks and attach the links here

**Comment by napetrov** [comment]
> @ethanglaser Thanks for the review! I guess @napetrov knows better when we can merge it? I will launch nightly job + benchmarks and attach the links here: nightly: https://intel-ci.intel.com/ef54c842-52a7-f191-b313-a4bf010d0e2e

once ready i would say?

**Comment by Alexandr-Solovev** [comment]
@Alexsandruss any other comments, it doesn't affect scikit-learn-intelex due to it still uses icc/icl as default

**Comment by homksei** [comment]
Jobs with ICX as a build compiler:

Linux: [intel-ci.intel.com/ef65521e-d0d2-f199-b313-a4bf010d0e2e](https://intel-ci.intel.com/ef65521e-d0d2-f199-b313-a4bf010d0e2e)
Windows: [intel-ci.intel.com/ef65e568-6c9f-f164-a693-a4bf010d0e2e](https://intel-ci.intel.com/ef65e568-6c9f-f164-a693-a4bf010d0e2e)

(builds passed)

File: .ci/scripts/describe_system.sh (hunk 1 above)
Comment (Alexandr-Solovev) [reply .ci/scripts/describe_system.sh:60]: https://github.com/intel/scikit-learn-intelex/pull/2019

**Review by samir-nasibli — APPROVED** [review]
Assuming green CI. 

@homksei could you please dismiss @Alexsandruss change request in case, since the comment is already addressed?

**Review by homksei — APPROVED** [review]

**Pull request merged**